File: src/chain-observers/classifiers/change-recovery-in-progress-classifier.ts
import { CollectorClassifierBase, TRegisterEvaluationContext } from "./collector-classifier-base";

export interface IAccountChangingRecovery {
  accountToRecover: string;
  recoveryAccount: string;
  effectiveOn: Date;
}

export interface IChangeRecoveryInProgressData {
  recoveringAccounts: Record<string, IAccountChangingRecovery>;
}

export interface IChangeRecoveryCollectorOptions {
  changeRecoveryAccount: string;
}

export class ChangeRecoveryInProgressClassifier extends CollectorClassifierBase<
  {},
  IChangeRecoveryInProgressData,
  void,
  void,
  IChangeRecoveryCollectorOptions
> {
  public static forOptions(options: IChangeRecoveryCollectorOptions): TRegisterEvaluationContext {
    return {
      class: this, // Intentionally using `this` to refer to the class prototype itself later - even though it is not a class **instance**
      options
    };
  }
}

File: python/workerbee/chain_observers/classifiers/change_recovery_in_progress_classifier.py
from __future__ import annotations

from typing import Any

from ..classifier_results import ChangeRecoveryInProgressData
from .collector_classifier_base import CollectorClassifierBase, TRegisterEvaluationContext


class ChangeRecoveryInProgressClassifier(CollectorClassifierBase[ChangeRecoveryInProgressData]):
    """IChangeRecoveryInProgressData:
      recovering_accounts: dict[str, IAccountChangingRecovery]

    IAccountChangingRecovery:
      account_to_recover: str
      recovery_account: str
      effective_on: datetime
    """

    @classmethod
    def for_options(cls, options: dict[str, Any]) -> TRegisterEvaluationContext:
        return {"class": cls, "options": options}
